folly/io/IOStreamBuf-inl.h

```diff
@@ -0,0 +1,264 @@
+
+namespace folly {
+
+template <typename CharT, typename Traits>
+IOStreamBuf<CharT,Traits>::IOStreamBuf(const folly::IOBuf* head)
+  : std::basic_streambuf<CharT,Traits>(),
+    head_(head),
+    gcur_(head) {
+  csetg(gcur_->data(), gcur_->data(), gcur_->tail());
+}
+
+template <typename CharT, typename Traits>
+void IOStreamBuf<CharT,Traits>::swap(IOStreamBuf<CharT,Traits>& rhs) {
+  std::basic_streambuf<CharT,Traits>::swap(rhs);
+  std::swap(head_, rhs.head_);
+  std::swap(gcur_, rhs.gcur_);
+}
+
+template <typename CharT, typename Traits>
+const typename IOStreamBuf<CharT,Traits>::pos_type
+IOStreamBuf<CharT,Traits>::badoff =
+  static_cast<typename IOStreamBuf<CharT,Traits>::pos_type>(
+    static_cast<typename IOStreamBuf<CharT,Traits>::off_type>(-1));
+
+// This is called either to rewind the get area (because gptr() == eback())
+// or to attempt to put back a non-matching character (which we disallow
+// on non-mutable IOBufs).
+template <typename CharT, typename Traits>
+typename IOStreamBuf<CharT,Traits>::int_type
+IOStreamBuf<CharT,Traits>::pbackfail(int_type c) {
+  if (this->gptr() != this->eback())
+    return traits_type::eof(); // trying to putback non-matching character
+
+  if (gcur_ == head_) {
+    // Already at beginning of first IOBuf
+    return traits_type::eof();
+  }
+
+  // Find the next preceding non-empty IOBuf, back to head_
+  // Ensure the object state is not modified unless an earlier input sequence
+  // can be found.
+  const IOBuf* prev = gcur_;
+  do {
+    prev = prev->prev();
+  } while (prev->length() == 0 && prev != head_);
+
+  // Check whether c matches potential *gptr() before updating pointers
+  if (!Traits::eq(c, Traits::to_int_type(prev->tail()[-1])))
+    return traits_type::eof();
+
+  gcur_ = prev;
+
+  csetg(gcur_->data(), gcur_->tail() - 1, gcur_->tail());
+
+  return traits_type::to_int_type(*this->gptr());
+}
+
+template <typename CharT, typename Traits>
+typename IOStreamBuf<CharT,Traits>::int_type
+IOStreamBuf<CharT,Traits>::underflow() {
+  // public methods only call underflow() when gptr() >= egptr()
+  // (but it's not an error to call underflow when gptr() < egptr())
+  if (UNLIKELY(this->gptr() < this->egptr()))
+    return traits_type::to_int_type(*this->gptr());
+
+  // Also handles non-chained
+  const IOBuf* next = gcur_->next();
+  if (next == head_)
+    return traits_type::eof();
+
+  gcur_ = next;
+  csetg(gcur_->data(), gcur_->data(), gcur_->tail());
+
+  return traits_type::to_int_type(*this->gptr());
+}
+
+template <typename CharT, typename Traits>
+typename IOStreamBuf<CharT,Traits>::pos_type
+IOStreamBuf<CharT,Traits>::current_position() const {
+  pos_type pos = 0;
+
+  for (const IOBuf* buf = head_; buf != gcur_; buf = buf->next())
+    pos += buf->length();
+
+  return pos + (this->gptr() - this->eback());
+}
+
+template <typename CharT, typename Traits>
+typename IOStreamBuf<CharT,Traits>::pos_type
+IOStreamBuf<CharT,Traits>::seekoff(off_type off,
```

Comment:
Via @yfeldblum:
> This is very long. At the very least, split it into three private member functions for the three possible values of `way`, since the implementations for the three code paths share no code.